feat: Allow environment variables to parameterise the application

hadesrofl opened the pull request:
# :sparkles: Features

* feat: Using Spectre to enable a CLI with help and options. Saving a config file is optional as well has re-running the update process and showing a live clock in the console for it.
* feat: Allow environment variables to parameterise the application
* feat: Includes a docker build for linux amd64 and arm64

## New Contributors
* @hadesrofl made their first contribution

hadesrofl committed:
feat: Allow environment variables to parameterise the application

# :sparkles: Features

* Allows to use environment variables to parameterise the application
* Includes a docker build for linux amd64 and arm64

DjNemas commented on README.md:
```diff
@@ -2,24 +2,45 @@
-Both API tokens can be generated in your CCP under "Stammdaten/>_API". (Keep them Secret!!!)
+Both API tokens can be generated in your CCP under "Stammdaten/>_API" or respectively "Master Data/>_API". (Keep them Secret!!!)
+
+## Alternatively use Environment variables
```
~~Please add a note that environment variables have higher priority than the config disk file.~~

DjNemas replied:
Oh, I found out that you are creating the AccountInformation.json with the values from the environment.
Please mention this behaviour in the Readme.md. Otherwise people will be confused.

hadesrofl replied:
Just read the email notification about the first comment and wanted to check if I missed it in the readme. So you are saying to additionally mention that the ENV variables will be stored in the AccountInformation.json file, right?

DjNemas replied:
Yes right.
The thing is that changes to the values in the environment variable only take effect when the .json is deleted, as the values are automatically saved when the file is created. 
However, the process is skipped if the .json file already exists and only the values from the .json are used.

The user must know this, otherwise he will be confused.

hadesrofl replied:
Yeah, you are right. It contains an update for that as well as log messaging to the wait timer (see https://github.com/DjNemas/NetCupDynDNS/pull/2/files#diff-767a0135e8dc52442f467c06ad569200d303e241a813f795234bd9acfd90714bR48), as that one is a bit odd for real logging and there is little value in a ticking clock like logging.

One thought for some other time might be to add a flag to enable storing the values in the json file. That way a user could decide to store credentials or not. If I run this as a script, it might be enough to add parameters to the call like you do with bash scripts etc.

One last thing, if you merge this in, it might be a good idea to add a tag (v2.3.0) and create a new release, so that a docker container will be available for others in this repository. As you might have seen the workflow is triggered by tags that match `vX.Y.Z`

EDIT: Just because I redeployed the container. Here is an example of the new log message for the "retry"

<img width="1142" height="209" alt="Bildschirmfoto 2025-08-03 um 17 04 12" src="https://github.com/user-attachments/assets/18f88261-a79a-482c-a7ae-19c5d7730c73" />

DjNemas replied:
Hey hey nooo don't remove my ticking clock. 😢 To me it's not odd, and was one of the core features to show. You could show it with your additional information. But please leave it ticking. 😏 And we don't need to show milliseconds like 01:00:00 min. 01:00 min. is enough tbh.

Also, I saw that you can set the timer via environment variable but not via .json. You can't ignore the main users of this app who are already using this.
AccountInformation.json is perhaps the wrong name for such settings. We could think about a separate Settings.json or something like that. But right now it's just a value that is really important as a setting. So for me it would be ok to store this in AccountInformation.json as well.

hadesrofl replied:
## Logging behaviour

The issue with ticking is that it looks like this 
<img width="1064" height="935" alt="Bildschirmfoto 2025-08-03 um 17 57 59" src="https://github.com/user-attachments/assets/7365321a-9aed-491b-ab2f-79ecae58dcdc" />
 or this 
<img width="1559" height="475" alt="Bildschirmfoto 2025-08-03 um 17 56 37" src="https://github.com/user-attachments/assets/8a5f3cb5-6347-4429-b0c2-4e9e97d3ca71" />

in log systems. So it only looks "good" if you run it in your shell but not let the console outputs be picked up by some log viewer. What we can do is to add a flag to that application so that ticking is only deactivated as a container (by hardcoding it in the docker file to disable it). 

## Config file
In terms of using flags to enable features: The same applies to the config file. Storing their private information in a file without encryption etc. should be rather opt-in than opt-out, I feel. Right now the application is exposing passwords and api keys essentially via that file. An even better way compared to env variables would be to use the args of the application to set the necessary values only, including a flag if these information should be stored in a non-sensitive and secure way. 

## Retry behaviour

I want to highlight that there should be better ways to do a re-run of your application by scheduling it in the OS by using cron jobs or another task scheduler rather than keeping the process alive and even actively ticking a clock on the console. I don't think it should be a never ending process as there is no reason to it but left it for now to not make too many changes to the original code base. For my container use case it is convenient but I could also enable a task schedule to start the container again, execute the application and shut down the container again. That would be even cleaner also in terms of resource management. :-)

## Wait Time message format

Regarding the 01:00:00 minutes, that was 60 minutes before in HH:mm:ss format. That is why I changed it to only show <MINUTES> minutes now.

Let me know what you prefer. If you feel the name change of the config file is a breaking change, there is also the option to move to v3.0.0 with these changes.

DjNemas replied:
**Log**
I see your problem with a log viewer. I developed the application back then with the aim of making it as easy to use as possible. In other words, run the binary and see Terminal/CMD.
Now that you have extended it with Docker and are also using it with logging UIs, we can customise it to text only.

**Security/Config File**
Regarding security.
The normal way is to generate a .json file and only then write your data to the file manualy. It is local on the computer, and there is no web server or similar service. So how would you ever get access to the file?

**Seperate Environment/Json**
The fact that you are now writing your data from the environment variables into the JSON comes from you. :D
I would use an additional argument to indicate that environment variables should be used without a .json being generated. Without the specification of the argument, the standard way with the generation of the .json should be used.

**Retry behaviour**
Solution for a cron job. Also specify an argument that does not run the application in a while.
This application is supposed to run on its own without additional systems. But if you like it that way, you can implement such a feature.

**Time**
Regarding the time string.
I thought it was like mine before with one minute. That confused me. But then it fits if it's hours.

hadesrofl replied:
Yeah, I see the history of the application and liked that it solves the problem already. For my use case it just needed to be something that can actually be run on arm and in a different environment (NAS). But given this transition into something that is not an actively running application on your shell, it lacks behaviour that is uncommon nowadays. Logging to the terminal/CMD is totally in line with the 12 factor app, but lacks keeping that concept in mind (using Console.Write breaks with the principle as well as waiting to hit a key to exit the application).

## Security

Malware, as always. And if you consider this a weak argument, think about other critical systems that are offline for these kind of reasons. We shouldn't neglect something that obvious just for the sake of comfort. :-)

If we make storing it in a file optional, and name that it is clear text and not recommended, it's up for users to do it or not. It's them trading comfort for something potential risky.

I am doing a quick re-work adding new options for configs and retry. You can then decide if you want to merge that or not. Otherwise I leave my fork open. Thanks for your initial implementation anyway! :-)

hadesrofl replied:
Now it's clearly a breaking change and should be tagged as v3.0.0. ;-)

Please have a look. The biggest changes

* The application has a command line interface with options and a help
* Storing into the configuration file is optional and has to be manually enabled. Auto delete if storing is disabled.
* The live clock in the console is also an option that can be enabled.
* Order of precedence
  * Config file
  * Command arguments
  * ENV variables

Of course updated README.

A new option I could think of that could now be easily added since we have a real CLI is for enabling a dry run feature which logs the parameters instead of running the update process, so that people can see what the application is using to update their DNS records before putting something in and breaking their DNS settings. :-)

DjNemas commented:
All right, I've checked the code and it looks good. I still have to test the changes and if everything works, the changes will be merged.

hadesrofl committed:
fix: Make wait timer logs in containers more readable

hadesrofl committed:
fix: Change the minute expression to total minutes for logging

hadesrofl committed:
feat: Make it a full fledged command line application and use options to drive features

Also adds some options for storing into a configuration file, keeping the process alive and re-run the update after a certain interval and showing it as a live ticking clock in the console.

hadesrofl committed:
fix(WebRequest): Use ipify to determine ip address (#5)

DjNemas commented:
Hey, sorry @hadesrofl for the long delay. I totally lost focus on this pull request. I'll check it out in the next few days (maybe this weekend), and if everything is fine, I'll merge it into master.

DjNemas commented:
@hadesrofl 
I upgraded the project to .NET 10, performed an extensive refactor, fixed numerous bugs and compiler warnings, and implemented a custom CLI API inspired by Spectre to avoid unnecessary external dependencies. 
I also opened a pull request with these changes including your commits: https://github.com/DjNemas/NetCupDynDNS/pull/5. 

Please review the updated README and test the application. :)

This pull request will be closed.

hadesrofl commented:
Glad the contribution was helpful. :-)
